californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java

```diff
@@ -583,10 +586,13 @@ public void setCustomExecutor() {
 	 * @param ctx the correlation information
 	 */
 	public void setCorrelationContext(final CorrelationContext ctx) {
-		correlationContext = ctx;
-		ExchangeObserver obs = this.observer;
-		if (obs != null) {
-			obs.contextEstablished(this);
+		if (correlationContext.compareAndSet(null, ctx)) {
+			ExchangeObserver obs = this.observer;
+			if (obs != null) {
+				obs.contextEstablished(this);
+			}
+		} else {
+			correlationContext.set(ctx);
```

Comment:
I agree. But this is not really my question ?
I suppose by this response you mean that this is out of scope. That's ok. 
But I re-ask the question here as nobody really answer in https://github.com/eclipse/californium/issues/311#issuecomment-306852749